src/shared/components/Dialog.js

```diff
@@ -0,0 +1,152 @@
+ * @prop {Children} children - The content of the dialog.
+ * @prop {import("preact/hooks").Ref<HTMLElement>} [initialFocus] -
+ *   Child element to focus when the dialog is rendered.
+ * @prop {Children} [buttons] -
+ *   Additional `Button` elements to display at the bottom of the dialog.
+ *   A "Cancel" button is added automatically if the `onCancel` prop is set.
+ * @prop {string} [contentClass] - e.g. <button>
+ * @prop {'dialog'|'alertdialog'} [role] - The aria role for the dialog (defaults to" dialog")
+ * @prop {string} title - The title of the dialog.
+ * @prop {() => any} [onCancel] -
+ *   A callback to invoke when the user cancels the dialog. If provided, a
+ *   "Cancel" button will be displayed.
+ * @prop {string} [cancelLabel] - Label for the cancel button
+ */
+
+/**
+ * HTML control that can be disabled.
+ *
+ * @typedef {HTMLElement & { disabled: boolean }} InputElement
+ */
+
+/**
+ * A modal dialog wrapper with a title. The wrapper sets initial focus to itself
+ * unless an element inside of it is specified with the `initialFocus` ref.
+ * Optional action buttons may be passed in with the `buttons` prop but the
+ * cancel button is automatically generated when the on `onCancel` function is
+ * passed.
+ *
+ * Canonical resources:
+ *
+ * https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html
+ * https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/alertdialog.html
+ *
+ * If the dialog's content, specified by the `children` prop, contains a paragraph
+ * (`<p>`) element, that element will be identified as the dialog's accessible
+ * description.
+ *
+ * @param {DialogProps} props
+ */
+export default function Dialog({
+  children,
+  contentClass,
+  initialFocus,
+  onCancel,
+  cancelLabel = 'Cancel',
+  role = 'dialog',
+  title,
+  buttons,
+}) {
+  const dialogTitleId = useUniqueId('dialog-title');
+  const dialogDescriptionId = useUniqueId('dialog-description');
+
+  const rootEl = useRef(/** @type {HTMLDivElement | null} */ (null));
+
+  useElementShouldClose(rootEl, true, () => {
+    if (onCancel) {
+      onCancel();
+    }
+  });
+
+  useEffect(() => {
+    const focusEl = /** @type {InputElement|undefined} */ (initialFocus?.current);
+    if (focusEl && !focusEl.disabled) {
+      focusEl.focus();
+    } else {
+      // Modern accessibility guidance is to focus the dialog itself rather than
+      // trying to be smart about focusing a particular control within the
+      // dialog. See resources above.
+      rootEl.current.focus();
+    }
+    // We only want to run this effect once when the dialog is mounted.
+    //
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  // If the content of the dialog contains a paragraph of text, mark it as the
+  // dialog's accessible description.
+  //
+  // A limitation of this approach is that it doesn't update if the dialog's
+  // content changes after the initial render.
+  useLayoutEffect(() => {
```

Comment:
> Do I understand correctly that there will only be an aria-describedby for this dialog if there is at least one <p> element inside of the children and it will be set to an ID reference to the first paragraph? 

That's correct. I'll revise the comment a little. The exact use of the accessible description will depend on the assistive technology. IIRC VoiceOver reads it out after reading the title.